### ISSUE TITLE
Paper covering additional tokens idea

### ISSUE THREAD

**Issue by jph00** [BlinkDL/RWKV-LM#27]
Hi there. You mention in the readme that you're interested in potentially adding some special tokens/markers to represent stuff like capitalisation. Just wanted to let you know we tried that in the ULMFiT paper, and it worked pretty well. You can read the details here: https://arxiv.org/abs/1801.06146 . We went beyond capitalisation and added some other tokens too.

Anyhoo this is just an FYI in case it's helpful to you.

**Comment by BlinkDL** [comment]
Good to know that :) Thank you.
got more details on "we add special tokens for upper-case words, elongation, and repetition"?

**Comment by jph00** [comment]
Hmm OK looking back at the paper it's only a brief mention : "In addition, to allow the language model to capture aspects that might be relevant for classification, we add special tokens for upper-case words, elongation, and repetition." I've looked around and found our original code from years ago - here it is:

```
re_rep = re.compile(r'(\S)(\1{3,})')
TK_REP = 'tk_rep'

def replace_rep(m):
    c,cc = m.groups()
    return f' {TK_REP} {len(cc)+1} {c} '

re_word_rep = re.compile(r'(\b\w+\W+)(\1{3,})')
TK_WREP = 'tk_wrep'

def replace_wrep(m):
    c,cc = m.groups()
    return f' {TK_WREP} {len(cc.split())+1} {c} '

TOK_UP,TOK_SENT,TOK_MIX = ' t_up ',' t_st ',' t_mx '

def do_caps(ss):
    res = []
    prev='.'
    re_word = re.compile('\w')
    re_nonsp = re.compile('\S')
    for s in re.findall(r'\w+|\W+', ss):
        res += ([TOK_UP,s.lower()] if (s.isupper() and (len(s)>2))
#                 else [TOK_SENT,s.lower()] if (s.istitle() and re_word.search(prev))
                else [s.lower()])
#         if re_nonsp.search(s): prev = s
    return ''.join(res)

def proc_text(s):
    s = re_rep.sub(replace_rep, s)
    s = re_word_rep.sub(replace_wrep, s)
    s = do_caps(s)
    s = re.sub(r'([/#])', r' \1 ', s)
    s = re.sub(' {2,}', ' ', s)
    return spacy_tok(s)
```

So it lowercases everything, and before a word with any uppercase letter it adds a special token for either "lowercase", "uppercase", or "mixed case". Also, for words that repeat 3+ times it replaces them with a single word, prefixed by the number of repeats, and a special repeat-word token, and ditto for letters that repeat 3+ times.

**Comment by jph00** [comment]
(Actually this particular version doesn't seem to have done the `TOK_SENT`, `TOK_MIX` ones - I guess that must have been in a different file.)

**Comment by BlinkDL** [comment]
Nice. I'd like to save token count, and the "repeat 3+ times" idea is like RLE compression.

**Comment by jph00** [comment]
Yes it is like RLE, except of course the purpose isn't compression, but to create tokens with clearer semantics to make it easier for the model to learn meaning
